### DISCUSSION TITLE
timedelta fields are always serialized into ISO8601 format if response model is specified

### DISCUSSION THREAD

**Discussion by b564518128** [fastapi/fastapi#10174] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from datetime import timedelta
from fastapi import FastAPI
from pydantic import BaseModel, ConfigDict


class TimeDeltaModel(BaseModel):
    duration: timedelta = 0

    model_config = ConfigDict(ser_json_timedelta="float")


app = FastAPI()


@app.get("/")
async def root() -> TimeDeltaModel:
    return TimeDeltaModel(duration=80)
```

### Description

In the code above, if the return type designation (or `response_model` setting) is set, the `timedelta` object will always be serialized in ISO8601 format. If the return type designation is not set, it will be serialized according to Pydantics settings.

### Operating System

Linux

### FastAPI Version

0.101.1

### Pydantic Version

2.2.0

### Python Version

Python 3.10.12

### Additional Context

I am trying to serve a web site that shows the time balance for a particular activity. I am trying to use `timedelta` for this field since I don't need to implement time conversion functions as of using `integer` type. However, I need it to be serialized into `float` so that my frontend javascript can directly load it as javascript is not supporting `timedelta` natively.

**Comment by Kludex** [comment, marked as answer]
This is a bug in Pydantic, which was solved on main already: https://github.com/pydantic/pydantic/pull/7269.

It will be available in the next Pydantic release.

**Reply by atulenko** [reply]
Hi,

Testing with the latest pydantic version which contains that PR still reproduces this issue using the example code in this PR. 

I have tested with:

fastapi == 0.103.1
pydantic == 2.4.2
python == 3.11.4

I unfortunately haven't had the time to dig into the code enough to say in what exactly is happening. At a high level it appears the 
`TypeAdapter` pydantic class used within the `ModelField` object created in the `create_response_field` code no longer respects the configuration options.

ex code:

```
from pydantic import BaseModel, ConfigDict, create_model
from datetime import timedelta
from pydantic import TypeAdapter
from typing_extensions import Annotated
from pydantic.fields import FieldInfo
from fastapi._compat import (
    Undefined,
)

class TimeDeltaModel(BaseModel):
    duration: timedelta = 0

    model_config = ConfigDict(ser_json_timedelta="float")

ex_time_delta = TimeDeltaModel(duration=80)

print(ex_time_delta.model_dump_json())
# '{"duration":80.0}'

field_info = FieldInfo(annotation=TimeDeltaModel, default=Undefined, alias=None)

ta = TypeAdapter(Annotated[field_info.annotation, field_info])

print(ta.dump_python(ex_time_delta, mode="json"))
# {'duration': 'PT80S'}
```

I am not sure if TypeAdaptor is intended to respect all of the ConfigDict settings like ser_json_timedelta (in which case this would probably be a bug on pydantic) or if there is a different way to pass this config into the TypeAdaptor (in which case maybe we could consider using the TypeAdaptor differently in the fastAPI code).

**Comment by skion** [comment]
Can confirm the problem described in the topic start still exists with:

    pydantic                                 2.5.1
    pydantic-core                            2.14.3       
    fastapi                                  0.104.1 

Same happens when using `response_model=TimeDeltaModel` instead of the return type annotation.

In addition, the referenced upstream issue https://github.com/pydantic/pydantic/pull/7269 does seem to have been fixed with the above package versions. Therefore I suspect the issues are distinct and this discussion should be unmarked as _answered_ @Kludex ?

**Reply by igoose1** [reply]
As a confusing result of this discussion, I'm reminding @Kludex to unmark the answer.

FastAPI 0.108.0 and pydantic 2.5.3 don't fix the issue and I see this behavior as a bug.

**Reply by morian** [reply]
I this I may have some clues with this issue on another discussion similar to this one in discussion #10334 :
https://github.com/tiangolo/fastapi/discussions/10334#discussioncomment-8017482
